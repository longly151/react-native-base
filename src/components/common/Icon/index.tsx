import { StyleProp, TextStyle, TouchableOpacity, TouchableOpacityProps } from 'react-native';
import React from 'react';

import { useTheme } from '@react-navigation/native';
import { IconProps as VectorIconProps } from 'react-native-vector-icons/Icon';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
import getIconType from './getIconType';
import IconSource from '@icons';
import { SvgProps } from 'react-native-svg';
import withTailwind from '@components/hoc/withTailwind';

interface ISvg {
  name: keyof typeof IconSource;
  type: 'svg';
}
export interface IIconProps extends VectorIconProps {
  className?: string;
  primary?: boolean;
}
export type IconProps = IIconProps &
  (
    | IDefault
    | ISvg
    | IMaterial
    | IMaterialCommunity
    | ISimpleLineIcon
    | IZocial
    | IOcticon
    | IFontAwesome
    | IFontAwesome5
    | IIonicon
    | IFoundation
    | IEvilicon
    | IEntypo
    | IAntdesign
    | IFeather
    | IFontisto
  ) &
  Omit<SvgProps, 'style' | 'color'> & { touchableProps?: TouchableOpacityProps };

const IconFC: React.FC<IconProps> = React.forwardRef((props, ref: any) => {
  const theme = useTheme();
  // @ts-ignore
  const { style: styleProp, type, onPress, touchableProps, ...otherProps } = props;
  const color = props.primary ? theme.colors.primary : theme.colors.text;
  const style: StyleProp<TextStyle> = [styleProp];
  let Component;
  if (type === 'svg') {
    Component = IconSource[props.name as keyof typeof IconSource];
  } else {
    Component = getIconType(type);
  }

  const IconComponent = (
    <Component testID="Icon" ref={ref} size={25} color={color} {...otherProps} style={style}>
      {props.children}
    </Component>
  );

  if (onPress) {
    return (
      <TouchableOpacity testID="IconTouchableOpacity" onPress={onPress} {...touchableProps}>
        {IconComponent}
      </TouchableOpacity>
    );
  }
  return IconComponent;
});

IconFC.defaultProps = {
  type: 'material-community',
};

export interface IconType extends MaterialIcon {}

const Icon = withTailwind(IconFC) as React.FC<IconProps & { ref?: React.LegacyRef<MaterialIcon> }>;

export default Icon;
