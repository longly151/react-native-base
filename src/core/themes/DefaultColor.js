module.exports = {
  dark: false,
  colors: {
    primary: '#219EE5',
    background: '#f2f2f7',
    card: '#ffffff',
    text: '#000000',
    border: '#d8d8d8',
    notification: '#ff3b30',
    grey0: '#393e42',
    grey1: '#43484d',
    grey2: '#5e6977',
    grey3: '#86939e',
    grey4: '#bdc6cf',
    grey5: '#e1e8ee',
    white: '#ffffff',
    black: '#242424',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
