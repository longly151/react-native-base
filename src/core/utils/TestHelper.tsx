import React, { useLayoutEffect } from 'react';
import { render, RenderAPI } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import store, { persistor, createStore } from '@configs/store';
import { LoadingScreen } from '@app';
import { NavigationContainer } from '@react-navigation/native';
import Navigation from './Navigation';
import MockAdapter from 'axios-mock-adapter';
import { Axios } from '@utils/Api';

import DefaultColor from '@themes/DefaultColor';
import { RootState, useAppDispatch, useAppSelector } from './Redux';
import _ from 'lodash';

import { renderHook } from '@testing-library/react-hooks';

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>;
    }
  : T;

/**
 * Login Handler
 */
export const successCredential = { email: 'admin@gmail.com', password: '123456' };
export const successLoginResponse = {
  data: {
    token:
      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwic3RhdHVzIjoiQUNUSVZFIiwicm9sZSI6IkFETUlOIiwicm9sZUlkIjoxLCJwZXJtaXNzaW9ucyI6WyJBTEwiXSwidGFibGUiOiJ1c2VycyIsImlkRm9yZWlnbktleSI6InVzZXJJZCIsImlhdCI6MTY1NTQ1NzU5OCwiZXhwIjoxNjYwNjQxNTk4fQ.ZA1dsxArHAjb60HzQgKbkinl49cf5OiDWASiE_Wr0SA',
    id: 1,
    email: 'admin@gmail.com',
    fullName: 'Admin',
    phone: '0327571918',
    avatar:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcRlUbAyS_643dq_B69jZAlPNW6_Xc7SLELY6SpRsc5OI2wHiiYG&usqp=CAU',
    role: {
      createdAt: '2021-04-19T05:58:23.094Z',
      updatedAt: '2021-04-19T05:58:23.094Z',
      deletedAt: null,
      id: 1,
      name: 'ADMIN',
      description: 'ADMIN',
      permissions: [
        {
          id: 1,
          name: 'ALL',
          description: null,
        },
      ],
    },
  },
};

export const failCredential = { email: 'client@gmail.com', password: '123456' };
export const failLoginResponse = {
  statusCode: 401,
  message: [
    {
      code: 'INCORRECT_EMAIL_PASSWORD',
      description: 'Incorrect email or password',
    },
  ],
  error: 'Unauthorized',
};

const auth = {
  loading: false,
  data: successLoginResponse.data,
  error: null,
};
const withAuthStore = createStore({ auth });

/**
 * Wrapper handler
 */

export type OtherRenderOptions = {
  preloadedState?: DeepPartial<RootState>;
  withAuth?: boolean;
  disableNavigationRef?: boolean;
};

const AppContainer = ({
  children,
  disableNavigationRef,
}: {
  children: any;
  disableNavigationRef?: boolean;
}) => {
  const [loginData] = useAppSelector(state => [state.auth.data]);

  const dispatch = useAppDispatch();

  useLayoutEffect(() => {
    const initApp = () => {
      if (loginData?.token) {
        global.token = loginData.token;
      }
      global.fn = {} as any;
      global.dispatch = dispatch;
    };
    initApp();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return (
    <NavigationContainer
      ref={disableNavigationRef ? null : Navigation.navigationRef}
      theme={DefaultColor}
    >
      {children}
    </NavigationContainer>
  );
};

const AppProviders = ({
  children,
  preloadedState,
  withAuth,
  disableNavigationRef,
}: {
  children: any;
} & OtherRenderOptions) => {
  const mergedPreloadedState: DeepPartial<RootState> = withAuth
    ? _.merge({ auth }, preloadedState as any)
    : preloadedState;

  return (
    // Create new store if preloadedState !== undefined
    <Provider
      store={preloadedState ? createStore(mergedPreloadedState) : withAuth ? withAuthStore : store}
    >
      <PersistGate loading={LoadingScreen()} persistor={persistor}>
        <AppContainer disableNavigationRef={disableNavigationRef}>{children}</AppContainer>
      </PersistGate>
    </Provider>
  );
};

const customRender = (
  component: React.ReactElement<any>,
  // options?: RenderOptions & OtherRenderOptions,
  options?: OtherRenderOptions,
): RenderAPI => {
  const CustomWrapper = options?.wrapper || React.Fragment;
  const Wrapper = options?.wrapper
    ? ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={options.disableNavigationRef}
        >
          <CustomWrapper>{children}</CustomWrapper>
        </AppProviders>
      )
    : ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={options?.disableNavigationRef}
        >
          {children}
        </AppProviders>
      );

  return render(component, { ...options, wrapper: Wrapper });
};

const customRenderHook: typeof renderHook = (callback, options) => {
  const disableNavigationRef =
    typeof options?.disableNavigationRef === 'undefined' ? true : options?.disableNavigationRef;
  const CustomWrapper: any = options?.wrapper || React.Fragment;
  const Wrapper = options?.wrapper
    ? ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={disableNavigationRef}
        >
          <CustomWrapper>{children}</CustomWrapper>
        </AppProviders>
      )
    : ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={disableNavigationRef}
        >
          {children}
        </AppProviders>
      );

  return renderHook(callback, { ...options, wrapper: Wrapper });
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };

export { customRenderHook as renderHook };

const mockAxios = new MockAdapter(Axios, { delayResponse: 200 });

export { mockAxios };
