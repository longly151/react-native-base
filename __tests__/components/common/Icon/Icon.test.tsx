import React from 'react';
import { fireEvent, render } from '@utils/TestHelper';
import Icon from '@components/common/Icon/index';
import DefaultColor from '@themes/DefaultColor';

const mockTheme = DefaultColor;
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useTheme: jest.fn().mockImplementation(() => mockTheme),
}));

test('should render correctly', () => {
  const { toJSON, getByTestId } = render(<Icon name="arrow-left" />);
  expect(toJSON()).toMatchSnapshot();
  expect(getByTestId('Icon')).toBeTruthy();
});

test('should have working props', () => {
  /**
   * Vector Icon
   */
  const onIconPress = jest.fn();

  const { getByTestId, rerender, toJSON } = render(
    <Icon
      primary
      name="arrow-left"
      onPress={onIconPress}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500"
    />,
  );

  expect(getByTestId('Icon')).toHaveStyle({
    color: mockTheme.colors.primary,
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
    backgroundColor: 'rgba(14, 165, 233, 1)',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,
  });

  fireEvent.press(getByTestId('Icon'));

  expect(onIconPress).toBeCalledTimes(1);

  /**
   * SVG Icon
   */
  rerender(<Icon type="svg" name="loginBackground" />);
  expect(JSON.stringify(toJSON())).toContain('SvgMock');
});
