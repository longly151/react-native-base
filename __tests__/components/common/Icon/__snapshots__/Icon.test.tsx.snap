// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Text
  allowFontScaling={false}
  selectable={false}
  style={
    Array [
      Object {
        "color": "#000000",
        "fontSize": 25,
      },
      Array [
        Array [
          Object {},
          undefined,
        ],
      ],
      Object {
        "fontFamily": "Material Icons",
        "fontStyle": "normal",
        "fontWeight": "normal",
      },
      Object {},
    ]
  }
  testID="Icon"
>
  î—ž
</Text>
`;
